### PR TITLE
feat: enhance stableswap to easily support use in subpools

### PR DESCRIPTION
**Pull request by enthusiastmartin**
This PR refactors the stableswap pallet to be easily integrated into omnipool-subpools.

Mainly, it exposes necessary interface to use this pallet as dependency in another one. 

Math or pool logic has not changed.

### CODE REVIEW

File: stableswap/src/lib.rs [hunk 1]
```diff
@@ -716,4 +698,174 @@ impl<T: Config> Pallet<T> {
...
+        for asset in pool.assets.iter() {
+            ensure!(T::AssetRegistry::exists(*asset), Error::<T>::AssetNotRegistered);
+        }
+
+        Pools::<T>::insert(share_asset, pool);
+
+        Ok(share_asset)
+    }
+
+    pub fn add_asset_to_existing_pool(pool_id: T::AssetId, asset_id: T::AssetId) -> DispatchResult {
```
Comment (Roznovjak) [inline stableswap/src/lib.rs:747]: #[require_transactional]

File: stableswap/src/lib.rs [hunk 2]
```diff
@@ -716,4 +698,174 @@ impl<T: Config> Pallet<T> {
...
+                    let balance = T::Currency::free_balance(*asset, &orig_account);
+                    T::Currency::transfer(*asset, &orig_account, &pool_account, balance)?;
+                }
+            }
+
+            Ok(())
+        })
+    }
+
+    pub fn move_liquidity_to_pool(
```
Comment (Roznovjak) [inline stableswap/src/lib.rs:778]: #[require_transactional]

File: stableswap/src/tests/creation.rs [hunk 3]
```diff
@@ -107,27 +110,78 @@ fn create_pool_should_store_assets_correctly_when_input_is_not_sorted() {
...
+                Error::<Test>::ShareAssetInPoolAssets
+            );
+        });
+}
+
 #[test]
 fn create_pool_should_fail_when_asset_is_not_registered() {
     ExtBuilder::default()
         .with_registered_asset("one".as_bytes().to_vec(), 1000)
+        .with_registered_asset("pool".as_bytes().to_vec(), 100)
```
Comment (Roznovjak) [inline stableswap/src/tests/creation.rs:184]: `let pool_id: AssetId = 100;`

File: stableswap/src/tests/creation.rs [hunk 4]
```diff
@@ -73,8 +75,10 @@ fn create_pool_should_store_assets_correctly_when_input_is_not_sorted() {
     let asset_b: AssetId = 2;
     let asset_c: AssetId = 3;
     let asset_d: AssetId = 4;
+    let pool_id: AssetId = 100;
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/creation.rs:80]: `asset_a` and `asset_b`

File: stableswap/src/tests/update_pool.rs [hunk 5]
```diff
@@ -49,17 +50,18 @@ fn update_pool_should_work_when_all_parames_are_updated() {
 fn update_pool_should_work_when_only_amplification_is_updated() {
     let asset_a: AssetId = 1;
     let asset_b: AssetId = 2;
+    let pool_id: AssetId = 100;
 
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/update_pool.rs:56]:
```suggestion
        .with_endowed_accounts(vec![(ALICE, asset_a, 200 * ONE), (ALICE, asset_b, 200 * ONE)])
```

File: stableswap/src/tests/update_pool.rs [hunk 6]
```diff
@@ -90,17 +92,18 @@ fn update_pool_should_work_when_only_amplification_is_updated() {
 fn update_pool_should_work_when_only_trade_fee_is_updated() {
     let asset_a: AssetId = 1;
     let asset_b: AssetId = 2;
+    let pool_id: AssetId = 100;
 
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/update_pool.rs:98]:
```suggestion
        .with_endowed_accounts(vec![(ALICE, asset_a, 200 * ONE), (ALICE, asset_b, 200 * ONE)])
```

File: stableswap/src/tests/update_pool.rs [hunk 7]
```diff
@@ -131,17 +134,18 @@ fn update_pool_should_work_when_only_trade_fee_is_updated() {
 fn update_pool_should_work_when_only_withdraw_fee_is_updated() {
     let asset_a: AssetId = 1;
     let asset_b: AssetId = 2;
+    let pool_id: AssetId = 100;
 
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/update_pool.rs:140]:
```suggestion
        .with_endowed_accounts(vec![(ALICE, asset_a, 200 * ONE), (ALICE, asset_b, 200 * ONE)])
```

File: stableswap/src/tests/update_pool.rs [hunk 8]
```diff
@@ -172,17 +176,18 @@ fn update_pool_should_work_when_only_withdraw_fee_is_updated() {
 fn update_pool_should_work_when_only_fees_is_updated() {
     let asset_a: AssetId = 1;
     let asset_b: AssetId = 2;
+    let pool_id: AssetId = 100;
 
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/update_pool.rs:182]:
```suggestion
        .with_endowed_accounts(vec![(ALICE, asset_a, 200 * ONE), (ALICE, asset_b, 200 * ONE)])
```

File: stableswap/src/tests/update_pool.rs [hunk 9]
```diff
@@ -213,17 +218,18 @@ fn update_pool_should_work_when_only_fees_is_updated() {
 fn update_pool_should_fail_when_nothing_is_to_update() {
     let asset_a: AssetId = 1;
     let asset_b: AssetId = 2;
+    let pool_id: AssetId = 100;
 
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/update_pool.rs:224]:
```suggestion
        .with_endowed_accounts(vec![(ALICE, asset_a, 200 * ONE), (ALICE, asset_b, 200 * ONE)])
```

File: stableswap/src/tests/update_pool.rs [hunk 10]
```diff
@@ -271,17 +277,18 @@ fn update_pool_should_fail_when_pool_does_not_exists() {
 fn update_pool_should_fail_when_amplification_is_outside_allowed_range() {
     let asset_a: AssetId = 1;
     let asset_b: AssetId = 2;
+    let pool_id: AssetId = 100;
 
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/update_pool.rs:283]:
```suggestion
        .with_endowed_accounts(vec![(ALICE, asset_a, 200 * ONE), (ALICE, asset_b, 200 * ONE)])
```

File: stableswap/src/tests/update_pool.rs [hunk 11]
```diff
@@ -8,17 +8,18 @@ use sp_runtime::Permill;
 fn update_pool_should_work_when_all_parames_are_updated() {
     let asset_a: AssetId = 1;
     let asset_b: AssetId = 2;
+    let pool_id: AssetId = 100;
 
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
```
Comment (Roznovjak) [inline stableswap/src/tests/update_pool.rs:14]:
```suggestion
        .with_endowed_accounts(vec![(ALICE, asset_a, 200 * ONE), (ALICE, asset_b, 200 * ONE)])
```

File: stableswap/src/tests/creation.rs [hunk 12]
```diff
@@ -162,6 +218,7 @@ fn create_pool_should_fail_when_asset_is_not_registered() {
 fn create_pool_should_when_same_pool_already_exists() {
     ExtBuilder::default()
         .with_endowed_accounts(vec![(ALICE, 1, 200 * ONE), (ALICE, 2, 200 * ONE)])
+        .with_registered_asset("pool".as_bytes().to_vec(), 100)
         .with_registered_asset("one".as_bytes().to_vec(), 1)
         .with_registered_asset("two".as_bytes().to_vec(), 2)
```
Comment (Roznovjak) [inline stableswap/src/tests/creation.rs:223]: `let asset_a: AssetId = 1;`
`let asset_b: AssetId = 2;`
`let pool_id: AssetId = 100;`

File: stableswap/src/lib.rs [hunk 13]
```diff
@@ -716,4 +698,174 @@ impl<T: Config> Pallet<T> {
...
+        assets: &[AssetLiquidity<T::AssetId>],
+    ) -> Result<Balance, DispatchError> {
+        let pool = Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+        ensure!(assets.len() <= pool.assets.len(), Error::<T>::MaxAssetsExceeded);
+        let mut added_assets = BTreeMap::<T::AssetId, Balance>::new();
+        for asset in assets.iter() {
+            ensure!(
+                Self::is_asset_allowed(pool_id, asset.asset_id, Tradability::ADD_LIQUIDITY),
+                Error::<T>::NotAllowed
+            );
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply stableswap/src/lib.rs:819]: no, subpools relies on this.

File: stableswap/src/lib.rs [hunk 14]
```diff
@@ -268,7 +292,8 @@ pub mod pallet {
         /// initial liquidity.
         ///
         /// Parameters:
-        /// - `origin`: Must be T::PoolMasterOrigin
+        /// - `origin`: Must be T::AuthorityOrigin
+        /// - `share_asset`: Preregisted share asset identifier
```
Comment (Roznovjak) [inline stableswap/src/lib.rs:296]:
```suggestion
        /// - `share_asset`: Preregistered share asset identifier
```

File: stableswap/src/tests/add_liquidity.rs [hunk 15]
```diff
@@ -61,13 +62,15 @@ fn add_initial_liquidity_should_work_when_called_first_time() {
 
 #[test]
 fn add_initial_liquidity_should_fail_when_lp_has_insufficient_balance() {
+    let pool_id: AssetId = 100u32;
     ExtBuilder::default()
         .with_endowed_accounts(vec![
             (BOB, 1, 200 * ONE),
```
Comment (Roznovjak) [inline stableswap/src/tests/add_liquidity.rs:68]: I would create constants for these two asset IDs. Maybe also for the pool ID, it's up to you.

**Review by Roznovjak — APPROVED** [review]
check the latest comments and you can merge it

File: stableswap/Cargo.toml [hunk 16]
```diff
@@ -1,20 +1,22 @@
 [package]
 name = 'pallet-stableswap'
-version = '1.0.1'
+version = '1.0.0'
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply stableswap/Cargo.toml:3]: it is not used anywhere, hence i decided to reset it to v1

File: stableswap/src/lib.rs [hunk 17]
```diff
@@ -142,6 +141,12 @@ pub mod pallet {
     #[pallet::getter(fn pools)]
     pub type Pools<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, PoolInfo<T::AssetId>>;
 
+    /// Tradability state of pool assets.
+    #[pallet::storage]
+    #[pallet::getter(fn asset_tradability)]
+    pub type AssetTradability<T: Config> =
+        StorageDoubleMap<_, Blake2_128Concat, T::AssetId, Blake2_128Concat, T::AssetId, Tradability, ValueQuery>;
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply stableswap/src/lib.rs:148]: yes. default is to allow all.

**Pull request merged**